lib/ipam/ipam.go

```diff
@@ -478,30 +516,117 @@ func (c ipamClient) autoAssign(ctx context.Context, num int, handleID *string, a
-				ips = append(ips, newIPs...)
-				rem = num - len(ips)
-				break
+				s.datastoreRetryCount ++
 			}
+			s.datastoreRetryCount ++
 		}
+		return nil, false, errors.New("Max retries hit - excessive concurrent IPAM requests")
+	}
+
+	return nil, false, errors.New("failed to find or claim a block")
+}
+
+func (c ipamClient) autoAssign(ctx context.Context, num int, handleID *string, attrs map[string]string, requestedPools []net.IPNet, version int, host string, maxNumBlocks int, rsvdAttr *HostReservedAttr) ([]net.IPNet, error) {
+
+	// First, get the existing host-affine blocks.
+	logCtx := log.WithFields(log.Fields{"host": host})
+	if handleID != nil {
+		logCtx = logCtx.WithField("handle", *handleID)
+	}
+	logCtx.Info("Looking up existing affinities for host")
+	pools, affBlocks, err := c.prepareAffinityBlocksForHost(ctx, requestedPools, version, host, rsvdAttr)
+	if err != nil {
+		return nil, err
+	}
+
+	logCtx.Debugf("Found %d affine IPv%d blocks for host: %v", len(affBlocks), version, affBlocks)
+	ips := []net.IPNet{}
+
+	// Record how many blocks we own so we can check against the limit later.
+	numBlocksOwned := len(affBlocks)
+
+	config, err := c.GetIPAMConfig(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	s := &blockAssignState{
+		client:                c,
+		version:               version,
+		host:                  host,
+		pools:                 pools,
+		remainingAffineBlocks: affBlocks,
+		hostReservedAttr:      rsvdAttr,
+		allowNewClaim:         true,
+	}
+
+assignLoop:
+	// Allocate the IPs.
+	for len(ips) < num {
+		var b *model.KVPair
 
-		if retries == 0 {
-			return ips, errors.New("Max retries hit - excessive concurrent IPAM requests")
+		rem := num - len(ips)
+		if maxNumBlocks > 0 && numBlocksOwned >= maxNumBlocks {
+			s.allowNewClaim = false
+		}
+
+		b, newlyClaimed, err := s.findOrClaimBlock(ctx, 1)
+		if err != nil {
+			if _, ok := err.(noFreeBlocksError); ok {
+				break assignLoop
+			}
+			if errors.Is(err, ErrBlockLimit) {
+				log.Warnf("Unable to allocate a new IPAM block; host already has %v blocks but "+
+					"blocks per host limit is %v", numBlocksOwned, maxNumBlocks)
+				return ips, ErrBlockLimit
+			}
+			return nil, err
+		}
+
+		if newlyClaimed {
+			numBlocksOwned++
+		}
+
+	retryLoop:
+		// We have got a block b.
+		for i := 0; i < datastoreRetries; i++ {
+			newIPs, err := c.assignFromExistingBlock(ctx, b, rem, handleID, attrs, host, config.StrictAffinity)
+			if err != nil {
+				if _, ok := err.(cerrors.ErrorResourceUpdateConflict); ok {
+					log.WithError(err).Debug("CAS Error assigning from new block - retry")
+
+					// At this point, block b's Unallocated field has been reduced already.
+					// We should get the original block from datastore again.
+					blockCIDR := b.Key.(model.BlockKey).CIDR
+					b, err = c.blockReaderWriter.queryBlock(ctx, blockCIDR, "")
+					if err != nil {
+						logCtx.WithError(err).Warn("Failed to get block again after update conflict")
+						break retryLoop
```

Comment:
Actually, unless I'm reading this wrong, I think we could just remove the `retryLoop` tag entirely and the code would still work? We're never calling `break assignLoop` from within the inner loop, so `break retryLoop` is the same as just `break`

Reply:
Yeah. Code still works the same without tags. I'll get it removed.